path_dynamic_roles.go

```diff
@@ -0,0 +1,295 @@
+					Required:    true,
+				},
+				"username_template": {
+					Type:        framework.TypeString,
+					Description: "The template used to create a username",
+				},
+				"default_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Default TTL for dynamic credentials",
+				},
+				"max_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Max TTL a dynamic credential can be extended to",
+				},
+			},
+			ExistenceCheck: b.pathDynamicRoleExistenceCheck,
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.UpdateOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleCreateUpdate,
+				},
+				logical.CreateOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleCreateUpdate,
+				},
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleRead,
+				},
+				logical.DeleteOperation: &framework.PathOperation{
+					Callback:                    b.pathDynamicRoleDelete,
+					ForwardPerformanceStandby:   true,
+					ForwardPerformanceSecondary: true,
+				},
+			},
+			HelpSynopsis:    staticRoleHelpSynopsis,
+			HelpDescription: staticRoleHelpDescription,
+		},
+		// LIST role
+		{
+			Pattern: dynamicRolePath + "?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleList,
+				},
+			},
+			HelpSynopsis:    "List all the dynamic roles Vault is currently managing in OpenLDAP.",
+			HelpDescription: "List all the dynamic roles being managed by Vault.",
+		},
+		{
+			// Pattern: dynamicCredPath + framework.GenericNameRegex("name"),
+			Pattern: path.Join(dynamicCredPath, framework.MatchAllRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the dynamic role.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicCredsRead,
+				},
+			},
+			HelpSynopsis: "Request LDAP credentials for a dynamic role. These credentials are " +
+				"created within OpenLDAP when querying this endpoint.",
+			HelpDescription: "This path requests new LDAP credentials for a certain dynamic role. " +
+				"The credentials are created within OpenLDAP based on the creation_ldif specified " +
+				"within the dynamic role configuration.",
+		},
+	}
+}
+
+func secretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: secretCredsType,
+		Fields: map[string]*framework.FieldSchema{
+			"username": {
+				Type:        framework.TypeString,
+				Description: "Username of the generated account",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password to access the generated account",
+			},
+			"DNs": {
+				Type:        framework.TypeStringSlice,
```

Comment:
Is this correct, or should it be `"dn"` and of type `framework.TypeString`?

Reply:
At the time this was correct, however since we've changed it to allow for multiple statements this has been changed back to `DNs` as a string slice. Each element in the slice corresponds to the LDIF statement run.

Reply:
Should we change `"dn"` [in here](https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/21/files/d52a826c87f68ba2669287a3044b33da78cbcc60#diff-d5b65bfb287fcee080448ea1cbe9ba3f5af4f06b56b059bc2d0a59103f48a407R79) to `"DNs"` then?

Reply:
[It has already been changed](https://github.com/hashicorp/vault-plugin-secrets-openldap/pull/21/files#diff-d5b65bfb287fcee080448ea1cbe9ba3f5af4f06b56b059bc2d0a59103f48a407R76) - Your link is to a subset of changes that don't include this particular change.

Reply:
I'm not a fan of the capitalized field `"DNs"` since we tend to return snake cased field names, but I also don't have a better suggestion for the name at this point in time.

Reply:
I completely agree. Unfortunately we have a name conflict with `DNS` :(

We could expand to the full name: `distinguished_names`

Reply:
Updated to `distinguished_names`